File: py/symbolic/unreal.py
import json

from symbolic._lowlevel import lib, ffi
from symbolic._compat import range_type

from symbolic.utils import (
    RustObject,
    rustcall,
    decode_str,
    attached_refs,
    make_buffered_slice_reader,
)

__all__ = ["Unreal4Crash"]


class Unreal4Crash(RustObject):
    __dealloc_func__ = lib.symbolic_unreal4_crash_free

    @classmethod
    def from_bytes(cls, buffer):
        """Parses an Unreal Engine 4 crash"""
        buffer = ffi.from_buffer(buffer)
        rv = cls._from_objptr(
            rustcall(lib.symbolic_unreal4_crash_from_bytes, buffer, len(buffer))
        )
        attached_refs[rv] = buffer
        return rv

    def get_context(self):
        context_json = self._methodcall(lib.symbolic_unreal4_get_context)
        return json.loads(decode_str(context_json, free=True))

    def get_logs(self):
        logs_json = self._methodcall(lib.symbolic_unreal4_get_logs)
        return json.loads(decode_str(logs_json, free=True))

    @property
    def _file_count(self):
        """The count of files within the crash dump"""
        return self._methodcall(lib.symbolic_unreal4_crash_file_count)

    def _file_by_index(self, idx):
        """The file at the specified index within the dump"""
        rv = self._methodcall(lib.symbolic_unreal4_crash_file_by_index, idx)
        if rv == ffi.NULL:
            return None

        return Unreal4CrashFile._from_objptr(rv)

    def files(self):
        """Enumerate files within the UE4 crash"""
        for idx in range_type(self._file_count):
            yield self._file_by_index(idx)


class Unreal4CrashFile(RustObject):
    __dealloc_func__ = lib.symbolic_unreal4_file_free

    @property
    def name(self):
        """The file name."""
        name = self._methodcall(lib.symbolic_unreal4_file_name)
        return str(decode_str(name, free=True))

    @property
    def type(self):
        """The type of the file"""
        ty = self._methodcall(lib.symbolic_unreal4_file_type)
        return str(decode_str(ty, free=True))

    def open_stream(self):
        """Returns a stream to read files from the internal buffer."""
        len_out = ffi.new("uintptr_t *")
        rv = self._methodcall(lib.symbolic_unreal4_file_data, len_out)
        if rv == ffi.NULL:
            return None
        return make_buffered_slice_reader(ffi.buffer(rv, len_out[0]), self)

    @property
    def size(self):
        """Returns the size of the file in bytes."""
        len_out = ffi.new("uintptr_t *")
        self._methodcall(lib.symbolic_unreal4_file_data, len_out)
        return len_out[0]

File: symbolic-cabi/src/unreal.rs
use std::os::raw::c_char;
use std::slice;

use symbolic::unreal::{Unreal4Crash, Unreal4File};

use crate::core::SymbolicStr;
use crate::utils::ForeignObject;

/// An Unreal Engine 4 crash report.
pub struct SymbolicUnreal4Crash;

impl ForeignObject for SymbolicUnreal4Crash {
    type RustObject = Unreal4Crash;
}

/// A file contained in a SymbolicUnreal4Crash.
pub struct SymbolicUnreal4File;

impl ForeignObject for SymbolicUnreal4File {
    type RustObject = Unreal4File;
}

ffi_fn! {
    /// Parses an Unreal Engine 4 crash from the given buffer.
    unsafe fn symbolic_unreal4_crash_from_bytes(bytes: *const c_char, len: usize) -> Result<*mut SymbolicUnreal4Crash> {
        let slice = slice::from_raw_parts(bytes as *const _, len);
        let unreal = Unreal4Crash::parse(slice)?;
        Ok(SymbolicUnreal4Crash::from_rust(unreal))
    }
}

ffi_fn! {
    /// Frees an Unreal Engine 4 crash.
    unsafe fn symbolic_unreal4_crash_free(unreal: *mut SymbolicUnreal4Crash) {
        SymbolicUnreal4Crash::drop(unreal)
    }
}

ffi_fn! {
    /// Parses the Unreal Engine 4 context from a crash, returns JSON.
    unsafe fn symbolic_unreal4_get_context(
        unreal: *const SymbolicUnreal4Crash
    ) -> Result<SymbolicStr> {
        let context = SymbolicUnreal4Crash::as_rust(unreal).context()?;
        let json = serde_json::to_string(&context)?;
        Ok(json.into())
    }
}

ffi_fn! {
    /// Parses the Unreal Engine 4 logs from a crash, returns JSON.
    unsafe fn symbolic_unreal4_get_logs(unreal: *const SymbolicUnreal4Crash) -> Result<SymbolicStr> {
        let logs = SymbolicUnreal4Crash::as_rust(unreal).logs(100)?;
        let json = serde_json::to_string(&logs)?;
        Ok(json.into())
    }
}

ffi_fn! {
    /// Returns the number of files in the Unreal 4 crash.
    unsafe fn symbolic_unreal4_crash_file_count(
        unreal: *const SymbolicUnreal4Crash
    ) -> Result<usize> {
        Ok(SymbolicUnreal4Crash::as_rust(unreal).file_count())
    }
}

ffi_fn! {
    /// Returns file meta data of a file in the Unreal 4 crash.
    unsafe fn symbolic_unreal4_crash_file_by_index(
        unreal: *const SymbolicUnreal4Crash,
        index: usize,
    ) -> Result<*mut SymbolicUnreal4File> {
        Ok(match SymbolicUnreal4Crash::as_rust(unreal).file_by_index(index) {
            Some(file) => SymbolicUnreal4File::from_rust(file),
            None => std::ptr::null_mut()
        })
    }
}

ffi_fn! {
    /// Returns the file name of a file in the Unreal 4 crash.
    unsafe fn symbolic_unreal4_file_name(
        file: *const SymbolicUnreal4File
    ) -> Result<SymbolicStr> {
        Ok(SymbolicUnreal4File::as_rust(file).name().into())
    }
}

ffi_fn! {
    /// Returns the file type of a file in the Unreal 4 crash.
    unsafe fn symbolic_unreal4_file_type(
        file: *const SymbolicUnreal4File
    ) -> Result<SymbolicStr> {
        Ok(SymbolicUnreal4File::as_rust(file).ty().name().into())
    }
}

ffi_fn! {
    /// Returns the file contents of a file in the Unreal 4 crash.
    unsafe fn symbolic_unreal4_file_data(
        file: *const SymbolicUnreal4File,
        len: *mut usize,
    ) -> Result<*const u8> {
        let data = SymbolicUnreal4File::as_rust(file).data();

        if !len.is_null() {
            *len = data.len();
        }

        Ok(data.as_ptr())
    }
}

ffi_fn! {
    /// Frees an Unreal Engine 4 file.
    unsafe fn symbolic_unreal4_file_free(file: *mut SymbolicUnreal4File) {
        SymbolicUnreal4File::drop(file)
    }
}
